smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/model/NeedsReaderWriterTransformer.kt

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+package software.amazon.smithy.swift.codegen.model
+
+import software.amazon.smithy.model.Model
+import software.amazon.smithy.model.shapes.ServiceShape
+import software.amazon.smithy.model.shapes.ShapeType
+import software.amazon.smithy.model.shapes.StructureShape
+import software.amazon.smithy.model.traits.Trait
+import software.amazon.smithy.model.transform.ModelTransformer
+import software.amazon.smithy.swift.codegen.customtraits.NeedsReaderTrait
+import software.amazon.smithy.swift.codegen.customtraits.NeedsWriterTrait
+
+object NeedsReaderWriterTransformer {
+    fun transform(model: Model, service: ServiceShape): Model {
+        var modelToReturn = model
+
+        // Get all the input shapes, and tag their descendents with the NeedsWriterTrait
+        val inputShapes = model.getOperations(service).map { model.expectShape<StructureShape>(it.inputShape) }
+        inputShapes.forEach { modelToReturn = transform(modelToReturn, it, NeedsWriterTrait()) }
+
+        // Get all the output & error shapes, and tag their descendents with the NeedsReaderTrait
+        val outputShapes = model.getOperations(service).map { model.expectShape<StructureShape>(it.outputShape) }
+        val errorShapes = model.getErrorShapes(service)
+        (outputShapes + errorShapes).forEach { modelToReturn = transform(modelToReturn, it, NeedsReaderTrait()) }
+
+        return modelToReturn
+    }
+
+    private fun transform(model: Model, structure: StructureShape, trait: Trait): Model {
```

Comment:
This function is used recursively to traverse the model tree and add the trait (either `NeedsReader` or `NeedsWriter`) to structs and unions that descend from either an input (for `NeedsWriter`) or an output (for `NeedsReader`.)

Reply:
Shapes that already have the trait are excluded from the transformation to prevent endless loops when shapes have self-references.